mhkit/wave/contours.py

```diff
@@ -1521,4 +1522,331 @@ def _bivariate_KDE(x1, x2, bw, fit, nb_steps, Ndata_bivariate_KDE, kwargs):
     if return_fit:
         return x1_bivariate_KDE, x2_bivariate_KDE, fit
     return x1_bivariate_KDE, x2_bivariate_KDE
-    
+
+
+### Contour tools
+# TODO: Migrate from WDRT and use in environmental contour example
+def steepness():
+    """Calculate the breaking wave steepness curve to be plotted in a
+    Hs-Tp plot.
+
+    Can be used to modify calculated contours by setting the contour
+    Hs values to the breaking value if the contour Hs is larger than
+    the breaking Hs for any given period.
+    """
+    raise NotImplementedError()
+
+
+def outside_points():
+    """Return the (buoy) points outside a specific contour. """
+    raise NotImplementedError()
+
+
+def contour_integrator():
+    """Calculate the area enclosed by a contour. """
+    raise NotImplementedError()
+
+
+def bootstrap():
+    """Add confidence intervals (e.g. 95%) to calculated contours. """
+    raise NotImplementedError()
+
+
+def data_contour():
+    """Create a contour around the data. """
+    raise NotImplementedError()
+
+
+### Sampling
+def samples_full_seastate(x1, x2, points_per_interval, return_periods,
+                          sea_state_duration, method="PCA", bin_size=250):
+    """ Sample a sea-sate between contours of specified return periods.
+
+    This function is used for the full seastate approach for the
+    extreme load. See Coe et al. 2018 for more details. It was
+    originally part of WDRT.
+
+    Coe, R. G., Michelen, C., Eckert-Gallup, A., &
+    Sallaberry, C. (2018). Full long-term design response analysis of a
+    wave energy converter. Renewable Energy, 116, 356-366.
+
+    Parameters
+    ----------
+    x1: np.array
+        Component 1 data
+    x2: np.array
+        Component 2 data
+    points_per_interval : int
+        Number of sample points to be calculated per contour interval.
+    return_periods: np.array
+        Vector of return periods that define the contour intervals in
+        which samples will be taken. Values must be greater than zero
+        and must be in increasing order.
+    sea_state_duration : int or float
+        `x1` and `x2` sample rate (seconds)
+    method: string or list
+        Copula method to apply. Currently only 'PCA' is implemented.
+    bin_size : int
+        Number of data points in each bin
+
+    Returns
+    -------
+    Hs_Samples: np.array
+        Vector of Hs values for each sample point.
+    Te_Samples: np.array
+        Vector of Te values for each sample point.
+    weight_points: np.array
+        Vector of probabilistic weights for each sampling point
+        to be used in risk calculations.
+    """
+    if method != 'PCA':
```

Comment:
There are no checks on the types of data passed. These must be added.

Reply:
Done w/ @hivanov-nrel latest commit.